File: DD_c_plusplus/Examples/SetCoverInstance/SetCoverProblem.cpp


#include "SetCoverProblem.h"

#include <numeric>
#include <utility>

SetCoverProblem::SetCoverProblem(const SetCoverInstance& params, bool sort)
        : AbstractProblem(params.initial_state, params.variables)
{
    _raw_matrix = params.matrix_of_weight;

    // Identity default (overwritten by sort_variables when sort=true)
    dd_to_original_index.resize(params.variables.size());
    iota(dd_to_original_index.begin(), dd_to_original_index.end(), 0);

    if (sort) {
        apply_variable_order(sort_variables(params.variables));
    }

    dynamic_bitset<>* bs = convert_to_bitset(params.right_side_of_restrictions, _raw_matrix.size());
    this->right_side_of_restrictions = *bs;
    delete bs;

    for (const auto& row : _raw_matrix) {
        dynamic_bitset<>* bsr = convert_to_bitset(row, _raw_matrix[0].size());
        this->matrix_of_weight.push_back(*bsr);
        delete bsr;
    }

    create_transpose_matrix_of_weights(_raw_matrix);
    check_atributes(params.variables, params.initial_state);
    calculate_row_max_index();
}

vector<string> SetCoverProblem::sort_variables(const vector<pair<string, vector<int>>>& variables) {
    int n_vars = static_cast<int>(variables.size());
    int n_rows = static_cast<int>(_raw_matrix.size());

    vector<int> coverage(n_vars, 0);
    for (int j = 0; j < n_vars; ++j)
        for (int i = 0; i < n_rows; ++i)
            coverage[j] += _raw_matrix[i][j];

    vector<int> sort_indices(n_vars);
    iota(sort_indices.begin(), sort_indices.end(), 0);
    stable_sort(sort_indices.begin(), sort_indices.end(),
        [&coverage](int a, int b) { return coverage[a] > coverage[b]; });
    dd_to_original_index = sort_indices;

    vector<vector<int>> new_matrix(n_rows, vector<int>(n_vars));
    for (int i = 0; i < n_rows; ++i)
        for (int new_j = 0; new_j < n_vars; ++new_j)
            new_matrix[i][new_j] = _raw_matrix[i][sort_indices[new_j]];
    _raw_matrix = new_matrix;

    vector<string> sorted_names;
    sorted_names.reserve(n_vars);
    for (int idx : sort_indices)
        sorted_names.push_back(variables[idx].first);
    return sorted_names;
}

void SetCoverProblem::create_transpose_matrix_of_weights(vector<vector<int>>& vector_matrix_of_weight) {

    size_t rows = vector_matrix_of_weight.size();
    size_t var_num = vector_matrix_of_weight[0].size();
    matrix_of_weight_transpose.assign(var_num, dynamic_bitset<>(rows));

    for (size_t i = 0; i < rows; ++i) {
        for (size_t j = 0; j < var_num; ++j) {
            if(vector_matrix_of_weight[i][j] == 1){
                matrix_of_weight_transpose[j].set(i);
            }
        }
    }

}

dynamic_bitset<>* SetCoverProblem::convert_to_bitset(const vector<int>& vec, int size) {
    dynamic_bitset<>* bs = new dynamic_bitset<>(size);

    for (size_t i = 0; i < vec.size(); ++i) {
        (*bs)[i] = vec[i];
    }

    return bs;
}

void SetCoverProblem::check_atributes(const vector<pair<string, vector<int>>>& variables, dynamic_bitset<>* initial_state) {
    check_same_len_matrix_and_right_side(initial_state);
    check_same_len_rows_matrix_and_variables(variables);
}

void SetCoverProblem::check_same_len_matrix_and_right_side(dynamic_bitset<>* initial_state) {
    assert(matrix_of_weight.size() == right_side_of_restrictions.size() && "weights and capacity must have the same length");

    assert((initial_state->count() == right_side_of_restrictions.size() || initial_state->count() == 2 * right_side_of_restrictions.size()) &&
           "initial_state and capacity must have the same length");
}

void SetCoverProblem::check_same_len_rows_matrix_and_variables(const vector<pair<string, vector<int>>>& variables) {
    for (auto &row: matrix_of_weight) {
        assert(variables.size() == row.size() &&
               "weights and variables must have the same length");
    }
}

void SetCoverProblem::calculate_row_max_index() {
    max_indices.clear();
    for (size_t row = 0; row < matrix_of_weight.size(); ++row) {
        int max_index = -1;
        for (int col = matrix_of_weight[row].size() - 1; col >= 0; --col) {
            if (matrix_of_weight[row][col] == 1) {
                max_index = col + 1;
                break;
            }
        }
        max_indices.push_back(max_index);
    }
}

bool SetCoverProblem::transition_function(const dynamic_bitset<>* previous_state, const int variable_index, int variable_value, dynamic_bitset<>* scratch_state) const {
    if (variable_value == 0) {
        for (size_t row = 0; row < previous_state->size(); ++row) {
            if (max_indices[row] != -1 && variable_index >= max_indices[row] - 1) {
                if (previous_state->test(row))
                    return false;
            }
        }
        *scratch_state = *previous_state;
        return true;
    }
    *scratch_state = *previous_state;
    *scratch_state -= matrix_of_weight_transpose[variable_index];
    return true;
}

int SetCoverProblem::get_priority_for_discard_node(const dynamic_bitset<>* state) const {
    return state->count();
}

int SetCoverProblem::get_priority_for_merge_nodes(const int node_id, const dynamic_bitset<>* state) const {
    return state->count();
}

dynamic_bitset<>* SetCoverProblem::merge_operator(const dynamic_bitset<>* state_one, const dynamic_bitset<>* state_two) const {
    dynamic_bitset<>* merged_state = new dynamic_bitset<>((*state_one) & (*state_two));
    return merged_state;
}

string SetCoverProblem::get_state_as_string(const dynamic_bitset<>* state) const {
    string state_string;
    boost::to_string(*state, state_string);
    return state_string;
}

dynamic_bitset<>* SetCoverProblem::get_state_copy(const dynamic_bitset<>* state) const {
    dynamic_bitset<>* copy_bitset = new dynamic_bitset<>(*state);
    return copy_bitset;
}

dynamic_bitset<>* SetCoverProblem::get_final_state() {
    vector<int> final_state(matrix_of_weight.size(), 0);
    return convert_to_bitset(final_state, matrix_of_weight.size());
}

File: DD_python/Examples/SetCoverInstance/SetCoverProblem.py
from SourceCode.Problems.AbstractProblemClass import AbstractProblem
from bitarray import bitarray

class SetCoverProblem(AbstractProblem):

    def __init__(self, params: 'SetCoverStructure', sort: bool = True):
        '''
        Initialize the Set Cover Problem instance.

        Parameters:
        - params (SetCoverStructure): Instance of the set cover problem, containing:
            * initial_state: initial bitset state.
            * variables (list[tuple[str, list[int]]]): variable names and their domains.
            * matrix_of_weight (list[list[int]]): coverage matrix (rows=restrictions, cols=variables).
            * right_side_of_restrictions (list[int]): right-hand side of each restriction.
        - sort (bool): whether to sort variables by heuristic order.
        '''
        self._raw_matrix = params.matrix_of_weight
        super().__init__(params.initial_state, params.variables, sort)

        self.right_side_of_restrictions = self.convert_to_bitset(params.right_side_of_restrictions, len(params.matrix_of_weight))

        if not sort:
            self.matrix_of_weight = [
                self.convert_to_bitset(row, len(params.matrix_of_weight[0]))
                for row in params.matrix_of_weight
            ]
        self.matrix_of_weight_transpose = []

        self.check_atributes(params.variables, params.initial_state)
        self.create_transpose_matrix_of_weights(self.matrix_of_weight)
        self.calculate_row_max_index()

    def sort_variables(self, variables: list[tuple['Variable', 'Domain']]) -> list[str]:
        n_restrictions = len(self._raw_matrix)
        coverage = [
            sum(self._raw_matrix[row][i] for row in range(n_restrictions))
            for i in range(len(variables))
        ]
        self.dd_to_original_index = sorted(range(len(variables)), key=lambda i: -coverage[i])

        sorted_matrix = [[row[i] for i in self.dd_to_original_index] for row in self._raw_matrix]

        self.matrix_of_weight = [
            self.convert_to_bitset(row, len(sorted_matrix[0]))
            for row in sorted_matrix
        ]
    
        return [variables[i][0] for i in self.dd_to_original_index]

    def check_atributes(self, variables: list[tuple['Variable', 'Domain']], initial_state: 'State') -> None:
        '''
        Check the attributes of the problem.

        Parameters:
        - variables (list[tuple['Variable', 'Domain']]): List of tuples with the variables and their domains.
        - initial_state ('State'): Initial state of the problem.
        '''
        self.check_same_len_matrix_and_right_side(initial_state)
        self.check_same_len_rows_matrix_and_variables(variables)
    
    def check_same_len_matrix_and_right_side(self, initial_state: 'State') -> None:
        '''
        Check if the matrix of weight and the right side of restrictions have the same length.

        Parameters:
        - initial_state ('State'): Initial state of the problem.
        '''
        assert len(self.matrix_of_weight) == len(self.right_side_of_restrictions), "matrix_of_weight and right_side_of_restrictions must have the same length"
        assert len(initial_state) == len(self.right_side_of_restrictions) or len(initial_state) == 2*len(self.right_side_of_restrictions), "matrix_of_weight and right_side_of_restrictions must have the same length"
    
    def check_same_len_rows_matrix_and_variables(self, variables: list[tuple['Variable', 'Domain']]) -> None:
        '''
        Check if the rows of the matrix of weight and the right side of restrictions have the same length of the variables.

        Parameters:
        - variables (list[tuple['Variable', 'Domain']]): List of tuples with the variables and their domains.
        '''
        for row in range(len(self.matrix_of_weight)):
            assert len(self.matrix_of_weight[row]) == len(variables), "rows of matrix_of_weight and right_side_of_restrictions must have the same length of variables"

    def convert_to_bitset(self, vec, size):
        '''
        Convert a vector to a bitarray of a given size.

        Parameters:
        - vec (list[int]): Vector to convert. Values greater than 0 are set to True.
        - size (int): Size of the resulting bitarray.

        Returns:
        - bitarray: Bitarray representation of the vector.
        '''
        bs: bitarray = bitarray(size)
        bs.setall(False)
        for i, value in enumerate(vec):
            if value > 0:
                bs[i] = True
        return bs

    def create_transpose_matrix_of_weights(self, vector_matrix_of_weight: list[list[int]]) -> None:
        '''
        Build the transpose of the weight matrix as a list of bitarrays.

        Parameters:
        - vector_matrix_of_weight (list[list[int]]): Original weight matrix where each row
          represents a restriction and each column represents a variable.
        '''
        rows = len(vector_matrix_of_weight)
        var_num = len(vector_matrix_of_weight[0])

        self.matrix_of_weight_transpose = [bitarray(rows) for i in range(var_num) ]

        for i in range(rows):
            for j in range(var_num):
                if vector_matrix_of_weight[i][j] == 1:
                    self.matrix_of_weight_transpose[j][i] = True
    
    def calculate_row_max_index(self) -> None:
        '''
        Calculate the index of the last variable (column) set to 1 for each restriction (row).

        Populates self.max_indices where each entry is the 1-based index of the last
        True bit in the corresponding row of the weight matrix, or -1 if no bit is set.
        '''
        self.max_indices = []
        for row in self.matrix_of_weight:
            max_index = -1
            for col in range(len(row) - 1, -1, -1):
                if row[col]:
                    max_index = col + 1
                    break
            self.max_indices.append(max_index)

    def transition_function(self, previous_state: 'State', variable_index: int, variable_value: int, scratch_state: list) -> bool:
        if variable_value == 0:
            for row in range(len(previous_state)):
                if self.max_indices[row] != -1 and variable_index >= self.max_indices[row] - 1:
                    if previous_state[row]:
                        return False
            scratch_state[0][:] = previous_state
            return True
        scratch_state[0][:] = previous_state
        scratch_state[0] &= ~self.matrix_of_weight_transpose[variable_index]
        return True
    
    def get_priority_for_discard_node(self, state: 'State') -> int:
        '''
        Get the priority for discard a node.

        Parameters:
        - state ('State'): State of the problem.

        Returns:
        - int: Priority for discard the node.
        '''
        return state.count(True)
    
    def get_priority_for_merge_nodes(self, node_id: int, state: 'State') -> int:
        '''
        Get the priority for merge two nodes.

        Parameters:
        - id (int): Id of the node.

        Returns:
        - int: Priority for merge the nodes.
        '''
        return state.count()

    def merge_operator(self, state_one: 'State', state_two: 'State') -> 'State':
        '''
        Merge operator of the problem.

        Parameters:
        - state_one ('State'): First state.
        - state_two ('State'): Second state.

        Returns:
        - 'State': Merged state.
        '''
        return state_one & state_two
    
    def get_state_as_string(self, state: 'State') -> str:
        '''
        Get the state as a string.

        Parameters:
        - state ('State'): State of a node.

        Returns:
        - str: State as a string.
        '''
        return state.to01()
    
    def get_state_copy(self, state: 'State') -> 'State':
        '''
        Copy of a state.

        Parameters:
        - state(State): The state to copy. *READ IMPORTANT

        Returns:
        - State: A copy of the state.
        '''
        return state.copy()
        
    
    def get_final_state(self) -> 'State':
        '''
        Final state of the problem.

        Returns:
        - State: The final state of the problem.
        '''
        final_state = [0] * len(self.matrix_of_weight)
        return self.convert_to_bitset(final_state, len(self.matrix_of_weight))
